nodes/scene/evolver.py

```diff
@@ -0,0 +1,357 @@
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software Foundation,
+#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# ##### END GPL LICENSE BLOCK #####
+
+import random
+
+import bpy
+import bmesh
+import mathutils
+from mathutils import Vector, Matrix
+from math import radians
+from bpy.props import (
+    BoolProperty, FloatVectorProperty, StringProperty, EnumProperty, IntProperty, FloatProperty
+)
+from sverchok.node_tree import SverchCustomTreeNode, SverchCustomTree
+from sverchok.data_structure import dataCorrect, updateNode
+from sverchok.nodes.object_nodes.getsetprop import assign_data, types
+from mathutils.noise import seed_set, random
+import numpy as np
+import time
+from sverchok.core.update_system import process_from_nodes, make_tree_from_nodes, do_update
+from collections import namedtuple
+
+Gene = namedtuple('Gene', 'name g_type, min_n, max_n, range')
+# can't use a PointerProperty for type=bpy.data.node_group
+# https://blender.stackexchange.com/questions/2075/assign-datablock-to-custom-property
+evolver_mem = {}
+def is_valid_node(node, genotype_frame):
+    if genotype_frame == 'All' and node.bl_idname == "SvNumberNode":
+        return True
+    if node.parent and node.parent.name == genotype_frame and node.bl_idname == "SvNumberNode":
+        return True
+    return False
+
+def get_genes(target_tree, genotype_frame):
+    genes = []
+    for node in target_tree.nodes:
+        if is_valid_node(node, genotype_frame):
+            num_type = node.selected_mode
+            if node.selected_mode == "float":
+                min_n = node.float_min
+                max_n = node.float_max
+            else:
+                min_n = node.int_min
+                max_n = node.int_max
+            gene = Gene(name=node.name, g_type=num_type, min_n=min_n, max_n=max_n, range=max_n-min_n)
+            genes.append(gene)
+    return genes
+
+def initial_population(genes, init_pop_num):
+    population =[]
+    for i in range(init_pop_num):
+        agent = []
+        for g in genes:
+            a = g.min_n + random() * g.range
+            if g.g_type == 'int':
+                a = int(a)
+            agent.append(a)
+        population.append(agent)
+    return population
+
+
+def evaluate_fitness(population, genes, tree, update_list, node):
+
+    fitness = []
+    for agent in population:
+        tree.sv_process = False
+        for g, s in zip(genes, agent):
+            if g.g_type == 'float':
+                tree.nodes[g.name].float_= s
+            else:
+                tree.nodes[g.name].int_= s
+
+        tree.sv_process = True
```

Comment:
If you get an error here the `sv_process` attribute will remain unexpectedly false.

Reply:
What would you suggest? A try/except block or repeating 'sv_process=True' outside of the function?

Reply:
try/finally.

Reply:
Also you can use with statement. An example:

https://github.com/nortikin/sverchok/blob/9fb8060b96e733059324780a42b6bfb673b49694/core/events.py#L510-L519